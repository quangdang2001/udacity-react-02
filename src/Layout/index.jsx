export { default as LoginPage } from "./LoginPage/LoginPage";
export { default as PageTemplate } from "./BaseLayout/PageTemplate";
export { default as HomePage } from "./HomePage/HomePage";
export { default as QuestionPage } from "./QuestionPage/QuestionPage";
export { default as LeaderBoardPage } from "./LeaderBoardPage/LeaderBoardPage";
export { default as NewPage } from "./NewPage/NewPage";
export { default as NotFoundPage } from "./NotFoundPage/NotFoundPage";
