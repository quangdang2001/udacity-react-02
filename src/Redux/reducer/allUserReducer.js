import { _getUsers } from "../../Utils/_DATA";
import { fetchAllUser } from "../../Redux/action/allUserAction";

let allUserState = {
  allUser: [],
  pending: false,
};

export const asyncFetchAllUser = () => {
  return async (dispatch) => {
    let rawResult = await _getUsers();
    let result = Object.entries(rawResult).map((tupple) => tupple[1]);
    dispatch(fetchAllUser(result));
  };
};

export const allUserReducer = (state = allUserState, action) => {
  switch (action.type) {
    case "FETCH_ALL_USER":
      return {
        allUser: [...action.data],
        pending: false,
      };
    default:
      return state;
  }
};
