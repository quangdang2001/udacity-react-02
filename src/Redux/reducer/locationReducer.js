
let locationState = "/"

export const locationReducer = (state = locationState, action) => {
    switch (action.type) {
        case "CHANGE_LOCATION":
            return action.location
        default:
            return state
    }
}