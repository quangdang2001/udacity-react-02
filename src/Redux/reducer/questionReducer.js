import {
  fetchAllQuestion,
  preFetchAllQuestion,
} from "Redux/action/questionAction";
import { refreshData } from "Utils/actionUtil";
import { _getQuestions, _saveQuestion, _saveQuestionAnswer } from "Utils/_DATA";

let questionState = {
  allQuestion: [],
  newQuestions: [],
  doneQuestions: [],
  pending: false,
};

const isAnsweredQuestion = (answeredQuestionIdsArray, questionId) => {
  let check = answeredQuestionIdsArray.find(
    (ansQuestionId) => ansQuestionId === questionId
  );
  if (check) return true;
  else return false;
};

const sortQuestion = (questionA, questionB) => {
  return questionA?.timestamp < questionB?.timestamp ? 1 : -1;
};

export const asyncfetchAllQuestion = (answeredQuestions = {}) => {
  let answeredQuestionIdsArray =
    Object.entries(answeredQuestions).map((tupple) => tupple[0]) || [];
  return async (dispatch) => {
    dispatch(preFetchAllQuestion());
    let rawResult = await _getQuestions();
    let result = Object.entries(rawResult).map((tupple) => tupple[1]);
    let newQuestions = result
      .filter(
        (question) => !isAnsweredQuestion(answeredQuestionIdsArray, question.id)
      )
      .sort(sortQuestion);
    let doneQuestions = result
      .filter((question) =>
        isAnsweredQuestion(answeredQuestionIdsArray, question.id)
      )
      .sort(sortQuestion);
    dispatch(
      fetchAllQuestion({
        allQuestion: result,
        newQuestions: newQuestions,
        doneQuestions: doneQuestions,
      })
    );
  };
};

export const asyncAddQuestion = (question) => {
  return async (dispatch) => {
    dispatch(preFetchAllQuestion());
    await _saveQuestion(question);
    await refreshData(dispatch, question.author);
  };
};

export const asyncAddVotedQuestion = (votedData) => {
  return async (dispatch) => {
    dispatch(preFetchAllQuestion());
    await _saveQuestionAnswer(votedData);
    await refreshData(dispatch, votedData.authedUser);
  };
};

export const questionReducer = (state = questionState, action) => {
  switch (action.type) {
    case "FETCH_ALL_QUESTION":
      return {
        ...action.data,
        pending: false,
      };
    case "PRE_FETCH_ALL_QUESTION":
      return {
        allQuestion: [],
        newQuestions: [],
        doneQuestions: [],
        pending: true,
      };
    default:
      return state;
  }
};
