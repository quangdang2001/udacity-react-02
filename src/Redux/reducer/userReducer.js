let userState = {
  isLogin: false,
  info: {},
};
export const userReducer = (state = userState, action) => {
  switch (action.type) {
    case "LOGIN":
      return {
        isLogin: true,
        info: {
          ...action.payload,
        },
      };
    case "LOGOUT":
      return {
        isLogin: false,
        info: {},
      };
    case "FETCH_USER_INFO":
      return {
        ...state,
        info: { ...action.payload },
      };
    default:
      return state;
  }
};
