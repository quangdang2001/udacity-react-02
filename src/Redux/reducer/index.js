import { applyMiddleware, combineReducers, createStore } from "redux";
import { createLogger } from "redux-logger";
import thunk from "redux-thunk";
import { allUserReducer } from "./allUserReducer";
import { userReducer } from "./userReducer";
import { questionReducer } from "./questionReducer";

const middleware = [thunk, createLogger()];

const allReducers = combineReducers({
  users: allUserReducer,
  user: userReducer,
  questions: questionReducer,
});

export const store = createStore(allReducers, applyMiddleware(...middleware));
