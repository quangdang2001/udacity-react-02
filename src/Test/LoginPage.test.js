import React from "react";
import { render, fireEvent } from "@testing-library/react";
import { LoginPage } from "Layout";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import { store } from "Redux/reducer";

describe("TEST_LoginPage", () => {
  it("should show correct data when fill username", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <LoginPage />
        </BrowserRouter>
      </Provider>
    );
    const username = component.getByLabelText("Password:");
    const testData = "password";
    expect(username.value).toEqual("");
    fireEvent.change(username, { target: { value: testData } });
    expect(username.value).toEqual(testData);
  });

  it("should render correct1", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <LoginPage />
        </BrowserRouter>
      </Provider>
    );
    expect(component).toBeDefined();
  });

  it("should render correct2", () => {
    const component = render(
      <Provider store={store}>
        <BrowserRouter>
          <LoginPage />
        </BrowserRouter>
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });
});
