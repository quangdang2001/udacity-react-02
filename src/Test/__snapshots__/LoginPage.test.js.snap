// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TEST_LoginPage should render correct2 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="LoginPage__box"
      >
        <div
          class="LoginPage__title"
        >
          Employee Polls
        </div>
        <div
          class="LoginPage__imageBox"
        >
          <img
            alt="loginImage"
            class="LoginPage__image"
            src="LoginImage.png"
          />
        </div>
        <form
          class="LoginPage__formBox"
        >
          <div
            class="mb-3"
          >
            <label
              class="LoginPage__label form-label"
              for="username"
            >
              User:
            </label>
            <select
              class="form-select"
              id="username"
              placeholder="fill username..."
              required=""
            >
              <option
                disabled=""
                value=""
              >
                Select user
              </option>
            </select>
          </div>
          <div
            class="mb-3"
          >
            <label
              class="LoginPage__label form-label"
              for="password"
            >
              Password:
            </label>
            <input
              class="form-control"
              id="password"
              placeholder="Password"
              required=""
              type="password"
              value=""
            />
          </div>
          <button
            class="LoginPage__btnLogin btn btn-primary"
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="LoginPage__box"
    >
      <div
        class="LoginPage__title"
      >
        Employee Polls
      </div>
      <div
        class="LoginPage__imageBox"
      >
        <img
          alt="loginImage"
          class="LoginPage__image"
          src="LoginImage.png"
        />
      </div>
      <form
        class="LoginPage__formBox"
      >
        <div
          class="mb-3"
        >
          <label
            class="LoginPage__label form-label"
            for="username"
          >
            User:
          </label>
          <select
            class="form-select"
            id="username"
            placeholder="fill username..."
            required=""
          >
            <option
              disabled=""
              value=""
            >
              Select user
            </option>
          </select>
        </div>
        <div
          class="mb-3"
        >
          <label
            class="LoginPage__label form-label"
            for="password"
          >
            Password:
          </label>
          <input
            class="form-control"
            id="password"
            placeholder="Password"
            required=""
            type="password"
            value=""
          />
        </div>
        <button
          class="LoginPage__btnLogin btn btn-primary"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
